src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -1,59 +1,285 @@
+          case Some(Leaf(leafHash, leafTd)) if isBetterBranch(leafTd) =>
+            log.debug("Found a better chain, about to reorganise")
+            reorganiseChainFromQueue(leafHash) match {
+              case Right((oldBranch, newBranch)) =>
+                val totalDifficulties = newBranch.tail.foldRight(List(leafTd)) { (b, tds) =>
+                  (tds.head - b.header.difficulty) :: tds
+                }
+                ChainReorganised(oldBranch, newBranch, totalDifficulties)
+
+              case Left(error) =>
+                BlockImportFailed(s"Error while trying to reorganise chain: $error")
+            }
+
+          case _ =>
+            BlockEnqueued
+        }
+      }
+    }
+  }
+
+  /**
+    * Once a better branch was found this attempts to reorganise the chain
+    * @param queuedLeaf - a block hash that determines a new branch stored in the queue (newest block from the branch)
+    * @return [[BlockExecutionError]] if one of the blocks in the new branch failed to execute, otherwise:
+    *        (oldBranch, newBranch) as lists of blocks
+    */
+  private def reorganiseChainFromQueue(queuedLeaf: ByteString): Either[BlockExecutionError, (List[Block], List[Block])] = {
+    val newBranch = blockQueue.removeBranch(queuedLeaf)
+    val parent = newBranch.head.header.parentHash
+    val bestNumber = blockchain.getBestBlockNumber()
+
+    val staleBlocksWithReceiptsAndTDs = removeBlocksUntil(parent, bestNumber).reverse
+    val staleBlocks = staleBlocksWithReceiptsAndTDs.map(_._1)
+    val staleTd = (for (block <- staleBlocks) yield blockQueue.enqueueBlock(block))
+      .lastOption.flatten.map(_.totalDifficulty).getOrElse(0)
+
+    val (executedBlocks, maybeError) = executeBlocks(newBranch)
+    maybeError match {
+      case None =>
+        Right(staleBlocks, executedBlocks)
+
+      case Some(error) =>
+        // executed blocks were saved in the blockchain, we need to remove them...
+        if (executedBlocks.nonEmpty) {
+          removeBlocksUntil(executedBlocks.head.header.parentHash, executedBlocks.last.header.number)
+        }
+
+        // ... and bring back the old ones
+        staleBlocksWithReceiptsAndTDs.foreach { case (block, receipts, td) =>
+          blockchain.save(block)
+          blockchain.save(block.header.hash, receipts)
+          blockchain.save(block.header.hash, td)
+        }
+
+        executedBlocks.foreach(blockQueue.enqueueBlock(_, bestNumber))
+
+        newBranch.diff(executedBlocks).headOption.foreach { block =>
+          blockQueue.removeSubtree(block.header.hash)
+        }
+
+        Left(error)
+    }
+  }
+
+  /**
+    * Executes a list blocks, storing the results in the blockchain
+    * @param blocks block to be executed
+    * @return a list of blocks that were correctly executed and an optional [[BlockExecutionError]]
+    */
+  private def executeBlocks(blocks: List[Block]): (List[Block], Option[BlockExecutionError]) = {
```

Comment:
Yes, this one can be converted quite easily. Not that it's needed because the number of blocks is also bounded by `BlockQueue`, but I can do it if it makes anyone more comfortable (I realise there is common sense of alarm against non-tail recursive functions, which I do not share 😉 )

As for your question, yes, at least I'm not concerned 😉. Besides we're operating on a tree here, and recursive functions are a natural way of doing that.